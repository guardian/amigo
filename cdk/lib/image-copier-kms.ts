import type { GuStackProps } from "@guardian/cdk/lib/constructs/core";
import { GuStack } from "@guardian/cdk/lib/constructs/core";
import type { App } from "aws-cdk-lib";
import { CfnOutput, RemovalPolicy } from "aws-cdk-lib";
import { ArnPrincipal, Effect, PolicyDocument, PolicyStatement } from "aws-cdk-lib/aws-iam";
import { Key } from "aws-cdk-lib/aws-kms";

export class ImageCopierKMSKey extends GuStack {
  constructor(scope: App, id: string, props: GuStackProps) {
    super(scope, id, props);

    const key = new Key(this, "KmsKey", {
      removalPolicy: RemovalPolicy.RETAIN,
      enableKeyRotation: true,
      policy: new PolicyDocument({
        statements: [
          new PolicyStatement({
            sid: "Enable IAM User Permissions",
            effect: Effect.ALLOW,
            principals: [new ArnPrincipal(`arn:aws:iam::${this.account}:root`)],
            actions: ["kms:*"],
            resources: ["*"],
          }),
          new PolicyStatement({
            sid: "Allow use of the key",
            effect: Effect.ALLOW,
            principals: [
              new ArnPrincipal(
                `arn:aws:iam::${this.account}:role/aws-service-role/autoscaling.amazonaws.com/AWSServiceRoleForAutoScaling`
              ),
            ],
            actions: [
              "kms:Encrypt",
              "kms:Decrypt",
              "kms:ReEncrypt*",
              "kms:GenerateDataKey*",
              "kms:DescribeKey",
              "kms:CreateGrant",
            ],
            resources: ["*"],
          }),
          new PolicyStatement({
            sid: "Allow attachment of persistent resources",
            effect: Effect.ALLOW,
            principals: [
              new ArnPrincipal(
                `arn:aws:iam::${this.account}:role/aws-service-role/autoscaling.amazonaws.com/AWSServiceRoleForAutoScaling`
              ),
            ],
            actions: ["kms:CreateGrant"],
            resources: ["*"],
            conditions: {
              Bool: { "kms:GrantIsForAWSResource": true },
            },
          }),
        ],
      }),
    });

    this.overrideLogicalId(key, {
      reason: "to preserve key",
      logicalId: "KmsKey",
    });

    new CfnOutput(this, "AmigoImageCopierKey", {
      description: "Amigo image copier key arn",
      value: key.keyArn,
      exportName: "amigo-imagecopier-key",
    });
  }
}
