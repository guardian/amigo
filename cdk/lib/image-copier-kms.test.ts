import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { KMSKey } from './image-copier-kms';

describe('The kms key stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new KMSKey(app, 'IntegrationTest', {
			stack: 'cdk',
			stage: 'TEST',
		});
		const template = Template.fromStack(stack);
		expect(template.toJSON()).toMatchSnapshot();
	});
});
