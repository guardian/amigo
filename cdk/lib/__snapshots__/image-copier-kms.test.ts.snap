// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The kms key stack matches the snapshot 1`] = `
Object {
  "Outputs": Object {
    "AmigoImageCopierKey": Object {
      "Description": "Amigo image copier key arn",
      "Export": Object {
        "Name": "amigo-imagecopier-key",
      },
      "Value": Object {
        "Fn::GetAtt": Array [
          "KmsKey",
          "Arn",
        ],
      },
    },
  },
  "Resources": Object {
    "KmsKey": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "EnableKeyRotation": true,
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:aws:iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
              "Sid": "Enable IAM User Permissions",
            },
            Object {
              "Action": Array [
                "kms:Encrypt",
                "kms:Decrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
                "kms:DescribeKey",
                "kms:CreateGrant",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:aws:iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":role/aws-service-role/autoscaling.amazonaws.com/AWSServiceRoleForAutoScaling",
                    ],
                  ],
                },
              },
              "Resource": "*",
              "Sid": "Allow use of the key",
            },
            Object {
              "Action": "kms:CreateGrant",
              "Condition": Object {
                "Bool": Object {
                  "kms:GrantIsForAWSResource": true,
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:aws:iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":role/aws-service-role/autoscaling.amazonaws.com/AWSServiceRoleForAutoScaling",
                    ],
                  ],
                },
              },
              "Resource": "*",
              "Sid": "Allow attachment of persistent resources",
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/amigo",
          },
          Object {
            "Key": "Stack",
            "Value": "cdk",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
